resolwe/flow/models/entity.py

```diff
@@ -30,6 +58,59 @@ class Meta(BaseCollection.Meta):
     #: entity type
     type = models.CharField(max_length=100, db_index=True, null=True, blank=True)
 
+    #: duplication date and time
+    duplicated = models.DateTimeField(blank=True, null=True)
+
+    def is_duplicate(self):
+        """Return True if entity is a duplicate."""
+        return bool(self.duplicated)
+
+    def _duplicate(self, contributor, inherit_collection_membership):
+        """Duplicate (make a copy)."""
+        duplicate = Entity.objects.get(id=self.id)
+        duplicate.pk = None
+        duplicate.slug = None
+        name_max_length = Entity._meta.get_field('name').max_length
+        duplicate.name = 'Copy of {}'.format(self.name)[:name_max_length]
+        duplicate.contributor = contributor
+        duplicate.duplicated = datetime.datetime.now()
+        duplicate.save(force_insert=True)  # pylint: disable=protected-access
+
+        # TODO
+        # Create duplication data migration history entry.
+
+        assign_contributor_permissions(duplicate)
+
+        # Fields to inherit from original entity.
+        duplicate.created = self.created
+        duplicate.save()
+
+        # Duplicate entity's data objects.
+        # pylint: disable=no-member
+        orig_and_duplicated_data = self.data.all().duplicate(contributor, return_mapping=True)
+        orig_and_duplicated_data_ids = dict(
+            [[item['original'].id, item['derivative'].id] for item in orig_and_duplicated_data]
+        )
+        duplicate.data.add(*orig_and_duplicated_data_ids.values())
+
+        if inherit_collection_membership:
+            collections = get_objects_for_user(
+                contributor,
+                'add_collection',
+                self.collections.all()  # pylint: disable=no-member
+            )
+            for collection in collections:
+                collection.entity_set.add(duplicate)
+
+                # Add duplicated data to collection only if the original
```

Comment:
If an entity in the collection it implies that also all data from the entity should also be in the collection, so you can just add it.